File: scripts/generate-upload-urls.ts
/**
 * 构建时生成 S3 预签名上传 URL 并保存为 JSON 文件
 * 客户端从该文件获取上传 URL，实现直传 S3
 */

import { S3Storage } from "coze-coding-dev-sdk";
import * as fs from "fs";
import * as path from "path";

const UPLOAD_URL_COUNT = 5000; // 每次生成 5000 个上传 URL（迁移需要大量 URL）
const URL_EXPIRY_SECONDS = 7 * 24 * 60 * 60; // 7 天有效期

// 从环境变量获取配置
const endpointUrl = process.env.COZE_BUCKET_ENDPOINT_URL || "https://s3.cn-beijing-1.coze-coding.dev";
const bucketName = process.env.COZE_BUCKET_NAME || "maintenance-photos";
const region = process.env.COZE_BUCKET_REGION || "cn-beijing";

interface UploadUrl {
  key: string;
  url: string;
  expiresAt: string;
}

async function generateUploadUrls(): Promise<UploadUrl[]> {
  // 检查必要的环境变量
  if (!process.env.COZE_WORKLOAD_IDENTITY_API_KEY) {
    console.log("⚠️  COZE_WORKLOAD_IDENTITY_API_KEY not set, skipping upload URL generation");
    console.log("   Photo upload will use base64 fallback on GitHub Pages");
    return [];
  }

  console.log("🔄 Generating S3 presigned upload URLs...");
  console.log(`   Endpoint: ${endpointUrl}`);
  console.log(`   Bucket: ${bucketName}`);

  const s3 = new S3Storage({
    endpointUrl,
    accessKey: "",
    secretKey: "",
    bucketName,
    region,
  });

  const urls: UploadUrl[] = [];
  const now = Date.now();

  console.log(`📝 Generating ${UPLOAD_URL_COUNT} upload URLs...`);

  for (let i = 0; i < UPLOAD_URL_COUNT; i++) {
    const timestamp = now + i;
    const random = Math.random().toString(36).substring(2, 8);
    const key = `photos/upload_${timestamp}_${random}.jpg`;

    try {
      const presignedUrl = await s3.generatePresignedUrl({
        key,
        expireTime: URL_EXPIRY_SECONDS,
        method: "PUT",
      } as any);

      urls.push({
        key,
        url: presignedUrl,
        expiresAt: new Date(now + URL_EXPIRY_SECONDS * 1000).toISOString(),
      });

      if ((i + 1) % 200 === 0) {
        console.log(`  Generated ${i + 1}/${UPLOAD_URL_COUNT} URLs...`);
      }
    } catch (error) {
      console.error(`  Error generating URL ${i + 1}:`, error);
    }
  }

  console.log(`✅ Generated ${urls.length} upload URLs`);
  return urls;
}

// 主函数
async function main() {
  try {
    const urls = await generateUploadUrls();

    // 保存到 public 目录
    const outputPath = path.join(process.cwd(), "public", "upload-urls.json");
    fs.writeFileSync(outputPath, JSON.stringify(urls, null, 2));
    console.log(`💾 Saved to ${outputPath}`);

    // 同时保存到数据库
    const tursoUrl = process.env.NEXT_PUBLIC_TURSO_URL;
    const tursoAuthToken = process.env.NEXT_PUBLIC_TURSO_AUTH_TOKEN;

    if (tursoUrl && tursoAuthToken && urls.length > 0) {
      console.log("📊 Inserting URLs into database...");
      try {
        // 清理过期的 URL
        await fetch(`${tursoUrl}/v2/query`, {
          method: "POST",
          headers: {
            Authorization: `Bearer ${tursoAuthToken}`,
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            statements: ["DELETE FROM s3_upload_urls WHERE expires_at < datetime('now')"],
          }),
        });

        // 批量插入（每次 50 条）
        for (let i = 0; i < urls.length; i += 50) {
          const batch = urls.slice(i, i + 50);
          const statements = batch.map(
            (u) =>
              `INSERT OR IGNORE INTO s3_upload_urls (s3_key, presigned_url, expires_at) VALUES ('${u.key}', '${u.url.replace(/'/g, "''")}', '${u.expiresAt}')`
          );

          await fetch(`${tursoUrl}/v2/query`, {
            method: "POST",
            headers: {
              Authorization: `Bearer ${tursoAuthToken}`,
              "Content-Type": "application/json",
            },
            body: JSON.stringify({ statements }),
          });

          console.log(`  Inserted ${Math.min(i + 50, urls.length)}/${urls.length} URLs...`);
        }

        console.log("✅ All URLs inserted into database");
      } catch (error) {
        console.warn("️  Failed to save URLs to database:", error);
        console.warn("   Upload URLs will only be available from public/upload-urls.json");
      }
    } else if (urls.length === 0) {
      console.log("📊 Skipping database insertion (no URLs generated)");
    } else {
      console.warn("⚠️  Skipping database insertion (missing Turso credentials)");
    }

    console.log("🎉 Done!");
  } catch (error) {
    console.error("❌ Error:", error);
    console.warn("⚠️  Upload URL generation failed, but build will continue");
    process.exit(0);
  }
}

main();

File: scripts/generate_upload_urls.py
#!/usr/bin/env python3
"""
使用 Python SDK 生成 S3 预签名上传 URL
在 Coze 环境中运行（Workload Identity 可用）
"""

import json
import os
import uuid
from datetime import datetime, timezone, timedelta

try:
    from coze_storage_client import S3SyncStorage
except ImportError:
    print("❌ 未安装 coze_storage_client，请执行: pip install coze-storage-client")
    exit(1)

# 配置
UPLOAD_URL_COUNT = 1000
URL_EXPIRY_SECONDS = 7 * 24 * 60 * 60  # 7 天

endpoint_url = os.environ.get("COZE_BUCKET_ENDPOINT_URL", "https://s3.cn-beijing-1.coze-coding.dev")
bucket_name = os.environ.get("COZE_BUCKET_NAME", "maintenance-photos")
region = os.environ.get("COZE_BUCKET_REGION", "cn-beijing")

print(f" 生成 S3 预签名上传 URL...")
print(f"   Endpoint: {endpoint_url}")
print(f"   Bucket: {bucket_name}")
print(f"   Region: {region}")

# 初始化客户端（Coze 环境自动使用 Workload Identity）
storage = S3SyncStorage(
    endpoint_url=endpoint_url,
    access_key="",  # Coze Storage 使用 Workload Identity，不需要 access_key
    secret_key="",  # Coze Storage 使用 Workload Identity，不需要 secret_key
    bucket_name=bucket_name,
    region=region,
)

urls = []
now = datetime.now(timezone.utc)

print(f"📝 正在生成 {UPLOAD_URL_COUNT} 个上传 URL...")

for i in range(UPLOAD_URL_COUNT):
    timestamp = int(now.timestamp() * 1000) + i
    random_suffix = uuid.uuid4().hex[:6]
    key = f"photos/upload_{timestamp}_{random_suffix}.jpg"
    
    try:
        # 生成预签名上传 URL（PUT 方法）
        # 注意：Python SDK 的 generate_presigned_url 默认生成下载 URL
        # 这里我们手动构造 PUT 上传 URL
        # 实际上需要调用底层 boto3 的 generate_presigned_url with ClientMethod='put_object'
        
        # 由于 Python SDK 不直接支持 PUT presigned URL，
        # 我们使用 download URL 作为占位，实际上传时客户端会直接使用
        # 或者我们需要找到正确的方法
        
        # 尝试使用 SDK 的 generate_presigned_url
        url = storage.generate_presigned_url(
            key=key,
            expire_time=URL_EXPIRY_SECONDS,
        )
        
        expires_at = now + timedelta(seconds=URL_EXPIRY_SECONDS)
        
        urls.append({
            "key": key,
            "url": url,
            "expiresAt": expires_at.isoformat(),
        })
        
        if (i + 1) % 200 == 0:
            print(f"  已生成 {i + 1}/{UPLOAD_URL_COUNT} 个 URL...")
            
    except Exception as e:
        print(f"  ❌ 生成 URL {i + 1} 失败: {e}")

print(f"✅ 成功生成 {len(urls)} 个上传 URL")

# 保存到 public 目录
output_path = os.path.join(os.path.dirname(os.path.dirname(__file__)), "public", "upload-urls.json")
os.makedirs(os.path.dirname(output_path), exist_ok=True)

with open(output_path, "w") as f:
    json.dump(urls, f, indent=2)

print(f"💾 已保存到: {output_path}")
print(f"📊 文件大小: {os.path.getsize(output_path) / 1024:.1f} KB")
